File: apple/TrainTimePhone/Services/PendingRouteNotifier.swift
import CoreLocation
import Foundation
import UserNotifications

/// The scheduled reminder split into its parts, for the in-app readout. walkMin
/// is nil outside distance-aware mode (nothing to break out); bufferMin is the
/// user's chosen lead. Lets the chip colour the calculated walk time apart from
/// the fixed buffer.
struct NotifyPlan {
    let notifyTs: Int
    let walkMin: Int?
    let bufferMin: Int
}

/// Schedules the "your train leaves soon" heads-up for a queued shared
/// route. Port of android app notify/PendingRouteNotifier.kt, using a
/// calendar-triggered local notification (delivered by the OS even after the
/// app is killed). Tap deep-links traintime://resumeroute.
enum PendingRouteNotifier {
    // One reminder at a time, so a constant identifier: adding replaces any
    // existing pending request atomically (the analog of Android's
    // enqueueUniqueWork(REPLACE)) and cancel removes exactly it.
    private static let identifier = "pendingRoute"

    // Category carrying the "Send to Watch" action, registered by the AppDelegate.
    // Attached to the reminder only when a Garmin is paired, so the action surfaces
    // on the watch (over ANCS) exactly when it can do something.
    static let garminCategory = "PENDING_ROUTE_GARMIN"
    static let sendToWatchAction = "SEND_TO_WATCH"

    /// Saved-route lead in seconds, from the user's Settings choice (minutes).
    private static var savedLeadSec: Int {
        let minutes = UserDefaults.standard.integer(forKey: "routeReminderLeadMinutes")
        return (minutes > 0 ? minutes : 15) * 60
    }

    /// Next-connection lead in seconds, set independently of the saved-route
    /// lead. Defaults to 3 min when unset.
    private static var connectionLeadSec: Int {
        let minutes = UserDefaults.standard.integer(forKey: "connectionReminderLeadMinutes")
        return (minutes > 0 ? minutes : 3) * 60
    }

    private static var distanceAware: Bool {
        UserDefaults.standard.bool(forKey: "distanceAwareReminder")
    }

    /// Absolute epoch-second the reminder is set to fire for this route, using
    /// the same rule as schedule. For the in-app "notified in X min" readout.
    static func nextNotifyTs(for route: PendingRoute) -> Int? {
        let dist = userDistanceMeters(for: route)
        return PendingRouteLogic.notifyTs(
            route, savedLeadSec: savedLeadSec, connectionLeadSec: connectionLeadSec,
            userDistanceMeters: dist,
            walkSecondsOverride: dist != nil ? routedWalkSec(for: route) : nil)
    }

    /// The reminder split into walk + buffer for the chip and resume prompt. Uses
    /// the same distance and leads as schedule, so the readout matches what fires.
    /// walkMin is nil in static mode or for a connection leg (no walk component).
    static func nextNotifyPlan(for route: PendingRoute) -> NotifyPlan? {
        let dist = userDistanceMeters(for: route)
        // In distance-aware mode, prefer the routed walk the phone measured live
        // (same MKDirections basis as tracking) over the straight-line estimate,
        // so the chip and the lead match what tracking shows on arrival.
        let routedSec = dist != nil ? routedWalkSec(for: route) : nil
        guard let notifyTs = PendingRouteLogic.notifyTs(
            route, savedLeadSec: savedLeadSec, connectionLeadSec: connectionLeadSec,
            userDistanceMeters: dist, walkSecondsOverride: routedSec) else { return nil }
        let connection = PendingRouteLogic.isConnectionLeg(route)
        let walkMin: Int?
        if connection {
            walkMin = nil
        } else if let routedSec {
            walkMin = Int((Double(routedSec) / 60).rounded())
        } else if let dist {
            walkMin = Int(GeoUtils.walkMinutes(distanceMeters: dist).rounded())
        } else {
            walkMin = nil
        }
        let bufferSec = connection ? connectionLeadSec : savedLeadSec
        return NotifyPlan(notifyTs: notifyTs, walkMin: walkMin, bufferMin: bufferSec / 60)
    }

    /// The live routed walk to the current leg's origin, measured by the phone
    /// (`PhoneViewModel.updatePendingRouteWalk`) via MKDirections and stashed here.
    /// Nil unless routed distance is on, the stored value belongs to this route,
    /// and it's fresh, so the notifier falls back to the straight-line estimate.
    private static func routedWalkSec(for route: PendingRoute) -> Int? {
        let defaults = UserDefaults.standard
        guard defaults.bool(forKey: "useRoutedDistance"),
              defaults.string(forKey: "pendingWalkRouteId") == route.id else { return nil }
        let stamp = defaults.integer(forKey: "pendingWalkTs")
        guard stamp > 0, Int(Date().timeIntervalSince1970) - stamp < 900 else { return nil }
        let sec = defaults.integer(forKey: "pendingWalkSec")
        return sec > 0 ? sec : nil
    }

    /// Straight-line distance from the last known location to the current leg's
    /// origin, only in distance-aware mode. Nil (static lead) when off, no leg,
    /// or no stored coordinate.
    private static func userDistanceMeters(for route: PendingRoute) -> Double? {
        guard distanceAware, let leg = route.currentLeg,
              let originLat = leg.originLat, let originLon = leg.originLon else { return nil }
        let lat = UserDefaults.standard.double(forKey: "lastLat")
        let lon = UserDefaults.standard.double(forKey: "lastLon")
        guard lat != 0, lon != 0 else { return nil }
        return GeoUtils.haversineDistance(
            from: CLLocationCoordinate2D(latitude: lat, longitude: lon),
            to: CLLocationCoordinate2D(latitude: originLat, longitude: originLon))
    }

    /// Contextual ask the first time anything schedules a reminder. Every
    /// schedule call happens app-foreground (share deep link, leg toggle,
    /// tracking hand-off), so the system prompt is always legal. Denial is
    /// fine: the chip and resume prompt work without the reminder. iOS
    /// evaluates authorisation at delivery, so a request racing the schedule
    /// still delivers once granted.
    static func requestAuthorizationIfNeeded() {
        let center = UNUserNotificationCenter.current()
        center.getNotificationSettings { settings in
            guard settings.authorizationStatus == .notDetermined else { return }
            center.requestAuthorization(options: [.alert, .sound]) { _, _ in }
        }
    }

    static func schedule(_ route: PendingRoute, now: Int, fromLocationUpdate: Bool = false) {
        requestAuthorizationIfNeeded()
        let dist = userDistanceMeters(for: route)
        guard let notifyTs = PendingRouteLogic.notifyTs(
            route, savedLeadSec: savedLeadSec, connectionLeadSec: connectionLeadSec,
            userDistanceMeters: dist,
            walkSecondsOverride: dist != nil ? routedWalkSec(for: route) : nil
        ), let leg = route.currentLeg else { return cancel() }
        if notifyTs <= now {
            // Inside the window. On a (re)save that means the user just saw the
            // route, so clear any stale reminder. An SLC wake landing here must
            // not clobber anything: the previously scheduled request still
            // fires, and a reminder already delivered stays on the lock screen.
            if !fromLocationUpdate { cancel() }
            return
        }

        let line = "\(leg.category ?? "")\(leg.lineNumber ?? "")"
        let formatter = DateFormatter()
        formatter.dateFormat = "HH:mm"
        formatter.timeZone = TimeZone(identifier: "Europe/Zurich")
        let depTime = formatter.string(from: Date(timeIntervalSince1970: TimeInterval(leg.depTs)))

        let content = UNMutableNotificationContent()
        let displayLine = line.isEmpty ? String(localized: "Train") : line
        content.title = String(localized: "\(displayLine) to \(route.finalDestination)")
        content.body = String(localized: "Departs \(depTime) from \(leg.originName)")
        content.sound = .default
        content.userInfo = ["deepLink": "traintime://resumeroute"]
        // Offer "Send to Watch" only when a Garmin has actually connected here at
        // least once (sticky, cached on watch refresh). The handler re-checks the
        // live link before sending.
        if UserDefaults.standard.bool(forKey: "garminEverConnected") {
            content.categoryIdentifier = garminCategory
        }

        let fireDate = Date(timeIntervalSince1970: TimeInterval(notifyTs))
        let components = Calendar.current.dateComponents(
            [.year, .month, .day, .hour, .minute, .second], from: fireDate)
        let trigger = UNCalendarNotificationTrigger(dateMatching: components, repeats: false)
        // Constant identifier: replaces any prior pending reminder atomically.
        let request = UNNotificationRequest(identifier: identifier, content: content, trigger: trigger)
        UNUserNotificationCenter.current().add(request)
    }

    /// Fires in a few seconds so the user can confirm permission + delivery
    /// without waiting for a real departure. Own identifier so it never
    /// disturbs a scheduled reminder.
    static func sendTest() {
        notify(title: String(localized: "Test reminder"), body: String(localized: "Route reminders are working. This is a test."))
    }

    /// Posts a short-delay notification in the reminder channel. Used by the test
    /// buttons (plain + distance readout). Own identifier so it never disturbs a
    /// scheduled reminder.
    static func notify(title: String, body: String) {
        requestAuthorizationIfNeeded()
        let content = UNMutableNotificationContent()
        content.title = title
        content.body = body
        content.sound = .default
        let trigger = UNTimeIntervalNotificationTrigger(timeInterval: 3, repeats: false)
        let request = UNNotificationRequest(
            identifier: "\(identifier)-test", content: content, trigger: trigger)
        UNUserNotificationCenter.current().add(request)
    }

    static func cancel() {
        let center = UNUserNotificationCenter.current()
        center.removePendingNotificationRequests(withIdentifiers: [identifier])
        center.removeDeliveredNotifications(withIdentifiers: [identifier])
    }

    // MARK: - Approach alert (background-card sessions with no saved route)

    // A board-tap "track in the background" session has no scheduled route
    // reminder, so it schedules its own one-shot "time to leave" here. Own
    // identifier so it never collides with the route reminder above. Mirrors the
    // Android FGS maybeApproachAlert (a saved route keeps using its reminder).
    private static let approachIdentifier = "approachAlert"

    /// Schedule the distance-aware "time to leave" alert: fires walk + buffer
    /// before the effective departure. No-op (and clears any prior one) when the
    /// window has already passed at schedule time.
    static func scheduleApproachAlert(_ focused: FocusedDeparture, walkSeconds: Int, now: Int) {
        cancelApproachAlert()
        requestAuthorizationIfNeeded()
        let lead = walkSeconds + savedLeadSec
        let fireIn = focused.departureTimestamp - lead - now
        guard fireIn > 0 else { return }

        let formatter = DateFormatter()
        formatter.dateFormat = "HH:mm"
        formatter.timeZone = TimeZone(identifier: "Europe/Zurich")
        let time = formatter.string(from: Date(timeIntervalSince1970: TimeInterval(focused.departureTimestamp + focused.delay * 60)))
        let line = focused.lineNumber.isEmpty ? String(localized: "Train") : focused.lineNumber

        let content = UNMutableNotificationContent()
        content.title = String(localized: "Time to leave")
        content.body = "\(line) \(focused.destination) · \(time)"
        content.sound = .default
        let trigger = UNTimeIntervalNotificationTrigger(timeInterval: TimeInterval(fireIn), repeats: false)
        let request = UNNotificationRequest(identifier: approachIdentifier, content: content, trigger: trigger)
        UNUserNotificationCenter.current().add(request)
    }

    static func cancelApproachAlert() {
        let center = UNUserNotificationCenter.current()
        center.removePendingNotificationRequests(withIdentifiers: [approachIdentifier])
        center.removeDeliveredNotifications(withIdentifiers: [approachIdentifier])
    }
}

File: android/app/src/main/kotlin/com/evanjt/traintime/notify/PendingRouteNotifier.kt
package com.evanjt.traintime.notify

import android.app.NotificationChannel
import android.app.NotificationManager
import android.content.Context
import androidx.core.app.NotificationCompat
import androidx.work.ExistingWorkPolicy
import androidx.work.OneTimeWorkRequestBuilder
import androidx.work.WorkManager
import com.evanjt.traintime.R
import com.evanjt.traintime.data.model.PendingRoute
import com.evanjt.traintime.data.prefs.AppPrefs
import com.evanjt.traintime.data.prefs.PendingRouteStore
import com.evanjt.traintime.data.sbb.SharedRoute
import com.evanjt.traintime.domain.GeoUtils
import com.evanjt.traintime.domain.LocaleUtil
import com.evanjt.traintime.domain.PendingRouteLogic
import java.util.UUID
import java.util.concurrent.TimeUnit
import kotlin.math.roundToInt
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.runBlocking

// The scheduled reminder split into its parts, for the in-app readout. walkMin
// is null outside distance-aware mode (nothing to break out); bufferMin is the
// user's chosen lead. Distinct from the raw notifyTs so the chip can colour the
// calculated walk time apart from the fixed buffer.
data class NotifyPlan(
    val notifyTs: Long,
    val walkMin: Int?,
    val bufferMin: Int,
)

// Schedules the "your train leaves soon" heads-up for a queued shared route.
// WorkManager rather than an exact alarm: it persists across process death
// and reboot, and inexact timing is fine for a 15-minute lead.
object PendingRouteNotifier {
    const val CHANNEL_ID = "pending_route"
    const val NOTIF_ID = 2
    private const val WORK_NAME = "pending-route-notify"

    // Notifications render outside any activity, so the per-app language override
    // (AppCompatDelegate) may not reach here on pre-33 devices. Resolve strings
    // through a context wrapped in the stored language tag.
    private fun localised(context: Context): Context {
        val tag = runBlocking { AppPrefs(context).appLanguage.first() }
        return LocaleUtil.localised(context, tag)
    }

    fun ensureChannel(context: Context) {
        val manager = context.getSystemService(NotificationManager::class.java)
        manager.createNotificationChannel(
            NotificationChannel(
                CHANNEL_ID,
                localised(context).getString(R.string.channel_saved_routes),
                NotificationManager.IMPORTANCE_HIGH,
            ),
        )
    }

    // Persist a synthesised route as a saved reminder and schedule it, without a
    // running ViewModel. Used by the watch's remind-on-phone path, which can
    // arrive with the app closed. Replaces any existing saved route (0.6.0 keeps
    // one); the VM's pending-route collector refreshes the chip if the app is up.
    suspend fun saveAndSchedule(context: Context, route: SharedRoute, nowEpochSeconds: Long): Boolean {
        val index = route.targetRideLegIndex(nowEpochSeconds) ?: return false
        val pending = PendingRoute.from(
            route = route,
            targetLegIndex = index,
            id = UUID.randomUUID().toString(),
            createdTs = nowEpochSeconds,
            sourceUrl = null,
        ).copy(status = PendingRoute.STATUS_SAVED)
        PendingRouteStore(context).save(pending)
        schedule(context, pending, nowEpochSeconds)
        return true
    }

    suspend fun schedule(
        context: Context,
        route: PendingRoute,
        nowEpochSeconds: Long,
        fromLocationUpdate: Boolean = false,
    ) {
        val prefs = AppPrefs(context)
        val savedLead = prefs.routeReminderLeadMinutes.first() * 60L
        val connectionLead = prefs.connectionReminderLeadMinutes.first() * 60L
        val distance = userDistanceMeters(prefs, route)
        val notifyTs = PendingRouteLogic.notifyTs(route, savedLead, connectionLead, distance)
            ?: return cancel(context)
        val delay = notifyTs - nowEpochSeconds
        if (delay <= 0) {
            // Inside the window. On a (re)save that means the user just saw the
            // route, so clear any stale reminder. A background fix landing here
            // must not clobber anything: the previously scheduled work still
            // fires, and a reminder already in the shade stays there.
            if (!fromLocationUpdate) cancel(context)
            return
        }
        WorkManager.getInstance(context).enqueueUniqueWork(
            WORK_NAME,
            ExistingWorkPolicy.REPLACE,
            OneTimeWorkRequestBuilder<PendingRouteNotifyWorker>()
                .setInitialDelay(delay, TimeUnit.SECONDS)
                .build(),
        )
    }

    // Absolute epoch-second the reminder is set to fire for this route, using
    // the same rule as schedule (distance-aware or static). For the in-app
    // "notified in X min" readout. Null when the route has no viable leg.
    suspend fun nextNotifyTs(context: Context, route: PendingRoute): Long? {
        val prefs = AppPrefs(context)
        val savedLead = prefs.routeReminderLeadMinutes.first() * 60L
        val connectionLead = prefs.connectionReminderLeadMinutes.first() * 60L
        return PendingRouteLogic.notifyTs(route, savedLead, connectionLead, userDistanceMeters(prefs, route))
    }

    // The reminder split into walk + buffer for the chip and resume prompt. Uses
    // the same distance and leads as schedule, so the readout matches what fires.
    // walkMin is null in static mode or for a connection leg (no walk component).
    suspend fun nextNotifyPlan(context: Context, route: PendingRoute): NotifyPlan? {
        val prefs = AppPrefs(context)
        val savedLead = prefs.routeReminderLeadMinutes.first() * 60L
        val connectionLead = prefs.connectionReminderLeadMinutes.first() * 60L
        val distance = userDistanceMeters(prefs, route)
        val notifyTs = PendingRouteLogic.notifyTs(route, savedLead, connectionLead, distance) ?: return null
        val connection = PendingRouteLogic.isConnectionLeg(route)
        val walkMin = if (distance != null && !connection) GeoUtils.walkMinutes(distance).roundToInt() else null
        val bufferSec = if (connection) connectionLead else savedLead
        return NotifyPlan(notifyTs, walkMin, (bufferSec / 60).toInt())
    }

    // Straight-line distance from the last known location to the current leg's
    // origin, only in distance-aware mode. Null (static lead) when the feature
    // is off, no leg, or no stored coordinate.
    private suspend fun userDistanceMeters(prefs: AppPrefs, route: PendingRoute): Double? {
        if (!prefs.distanceAwareReminder.first()) return null
        val leg = route.currentLeg ?: return null
        val originLat = leg.originLat ?: return null
        val originLon = leg.originLon ?: return null
        val (userLat, userLon) = prefs.lastCoordinate() ?: return null
        return GeoUtils.haversineDistance(userLat, userLon, originLat, originLon)
    }

    // Fires immediately so the user can confirm permission + delivery without
    // waiting for a real departure. Same channel as the real reminder.
    fun sendTest(context: Context) {
        val ctx = localised(context)
        notifyNow(context, ctx.getString(R.string.test_reminder_title), ctx.getString(R.string.test_reminder_body))
    }

    // Posts an immediate notification in the reminder channel. Used by the test
    // buttons (plain + distance readout).
    fun notifyNow(context: Context, title: String, body: String) {
        ensureChannel(context)
        val notification = NotificationCompat.Builder(context, CHANNEL_ID)
            .setSmallIcon(android.R.drawable.ic_menu_mylocation)
            .setContentTitle(title)
            .setContentText(body)
            .setStyle(NotificationCompat.BigTextStyle().bigText(body))
            .setAutoCancel(true)
            .build()
        context.getSystemService(NotificationManager::class.java)
            .notify(NOTIF_ID, notification)
    }

    fun cancel(context: Context) {
        WorkManager.getInstance(context).cancelUniqueWork(WORK_NAME)
        context.getSystemService(NotificationManager::class.java).cancel(NOTIF_ID)
    }
}
